File: app/dtos/experience.py
from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel


class WeeklyXpBreakdownItem(BaseModel):
    kind: str
    count: int
    points: int


class WeeklyXpResponse(BaseModel):
    week_id: str
    total_points: int
    items: list[WeeklyXpBreakdownItem]


class LeaderboardEntryDto(BaseModel):
    user_id: UUID
    user_name: str
    points: int
    rank: int


class LeaderboardResponse(BaseModel):
    week_id: str
    entries: list[LeaderboardEntryDto]
    my_rank: int | None
    my_points: int

File: web/src/types/experience.ts
/* =========================================
   주간 활동량(EXP) + 리더보드 타입
   ========================================= */

export type XpKind =
  | "HEALTH_INPUT"
  | "HEALTH_VIEW"
  | "CHALLENGE_VERIFY"
  | "POST"
  | "COMMENT"
  | "QUIZ";

export interface WeeklyXpBreakdownItem {
  kind: XpKind;
  count: number;
  points: number;
}

export interface WeeklyXpResponse {
  week_id: string; // "2026-W20"
  total_points: number;
  items: WeeklyXpBreakdownItem[];
}

export interface LeaderboardEntry {
  user_id: string; /* UUID */
  user_name: string;
  points: number;
  rank: number;
}

export interface LeaderboardResponse {
  week_id: string;
  entries: LeaderboardEntry[];
  my_rank: number | null;
  my_points: number;
}
